indexing-service/src/main/java/io/druid/indexing/common/task/SinglePhaseParallelIndexTaskRunner.java

```diff
@@ -0,0 +1,624 @@
+      final Set<SegmentIdentifier> segmentsIdentifiers = segmentsMap
+          .values()
+          .stream()
+          .flatMap(report -> report.getSegments().stream())
+          .map(SegmentIdentifier::fromDataSegment)
+          .collect(Collectors.toSet());
+      if (usedSegmentChecker.findUsedSegments(segmentsIdentifiers)
+                            .equals(segmentsToPublish)) {
+        log.info("Our segments really do exist, awaiting handoff.");
+      } else {
+        throw new ISE("Failed to publish segments[%s]", segmentsToPublish);
+      }
+    }
+  }
+
+  private void submitNewTask(
+      TaskMonitor<ParallelIndexSubTask> taskMonitor,
+      ParallelIndexSubTaskSpec spec
+  )
+  {
+    log.info("Submit a new task for spec[%s] and inputSplit[%s]", spec.getId(), spec.getInputSplit());
+    final ListenableFuture<SubTaskCompleteEvent<ParallelIndexSubTask>> future = taskMonitor.submit(spec);
+    Futures.addCallback(
+        future,
+        new FutureCallback<SubTaskCompleteEvent<ParallelIndexSubTask>>()
+        {
+          @Override
+          public void onSuccess(SubTaskCompleteEvent<ParallelIndexSubTask> completeEvent)
+          {
+            // this callback is called if a task completed wheter it succeeded or not.
+            taskCompleteEvents.offer(completeEvent);
+          }
+
+          @Override
+          public void onFailure(Throwable t)
+          {
+            // this callback is called only when there were some problems in TaskMonitor.
+            log.error(t, "Error while running a task for subTaskSpec[%s]", spec);
+            taskCompleteEvents.offer(SubTaskCompleteEvent.fail(spec, t));
+          }
+        }
+    );
+  }
+
+  @VisibleForTesting
+  int getAndIncrementNextSpecId()
+  {
+    return nextSpecId++;
+  }
+
+  @VisibleForTesting
+  Stream<ParallelIndexSubTaskSpec> subTaskSpecIterator() throws IOException
+  {
+    return baseFirehoseFactory.getSplits().map(this::newTaskSpec);
+  }
+
+  @VisibleForTesting
+  ParallelIndexSubTaskSpec newTaskSpec(InputSplit split)
+  {
+    return new ParallelIndexSubTaskSpec(
+        taskId + "_" + getAndIncrementNextSpecId(),
+        groupId,
+        taskId,
+        new ParallelIndexIngestionSpec(
+            ingestionSchema.getDataSchema(),
+            new ParallelIndexIOConfig(
+                baseFirehoseFactory.withSplit(split),
+                ingestionSchema.getIOConfig().isAppendToExisting()
+            ),
+            ingestionSchema.getTuningConfig()
+        ),
+        context,
+        split
+    );
+  }
+
+  private static List<InputSplit> getSplitsIfSplittable(FirehoseFactory firehoseFactory) throws IOException
+  {
+    if (firehoseFactory instanceof FiniteFirehoseFactory) {
+      final FiniteFirehoseFactory<?, ?> finiteFirehoseFactory = (FiniteFirehoseFactory) firehoseFactory;
+      return finiteFirehoseFactory.getSplits().collect(Collectors.toList());
```

Comment:
do you need to materialize the list? or can returning the `Stream` work?

Reply:
Yes, it should materialize because the returned value is only used for logging.